src/Illuminate/Collections/Collection.php

```diff
@@ -555,18 +555,37 @@ public function groupBy($groupBy, $preserveKeys = false)
      */
     public function keyBy($keyBy)
     {
-        $keyBy = $this->valueRetriever($keyBy);
-
         $results = [];
+        $items = $this->items;
 
-        foreach ($this->items as $key => $item) {
-            $resolvedKey = $keyBy($item, $key);
+        if (is_callable($keyBy) && ! is_string($keyBy)) {
```

Comment:
Why not use `$this->useAsCallable($value)` here?

Also, wouldn't other methods benefit from this if this added to the `valueRetriever` method?

Reply:
Also, wouldn't other methods benefit from this if this whole block was added to the `valueRetriever` method?